spec/javascripts/NewService/ServiceDiscoveryForm.spec.jsx

```diff
@@ -33,3 +35,40 @@ it('should render `ServiceDiscoveryListItems` child', () => {
   const wrapper = mount(<ServiceDiscoveryForm {...props}/>)
   expect(wrapper.find(ServiceDiscoveryListItems).exists()).toEqual(true)
 })
+
+describe('fetchProjects', () => {
+  const fetch = jest.spyOn(utils, 'fetchData')
+
+  afterEach(() => {
+    fetch.mockClear()
+  })
+
+  it('should render an error when fetching projects is unsuccessful', done => {
+    const msg = 'Something went wrong'
+    fetch.mockImplementation(url => { throw new Error(msg) })
+
+    const wrapper = mount(<ServiceDiscoveryForm {...props}/>)
+
+    expect(wrapper.find(ErrorMessage).exists()).toBe(true)
+    expect(wrapper.find(ErrorMessage).text()).toContain(msg)
+
+    setImmediate(done)
+  })
+
+  it('should fetch projects when first redendered', done => {
+    const projects = [{ name: 'project_00' }]
+    fetch.mockImplementation(url => projects)
+
+    const setState = jest.fn(val => {
+      expect(val).toEqual(projects)
+      done()
+    })
+    const useState = jest.spyOn(React, 'useState')
```

Comment:
I don't see a lot of value mocking React.useState just to assert it was called... And specially on init, not when it was expected to be called after a change or with specific values

Reply:
Rather than mocked it is _spied_, which is the only way [I know] to assert the projects are fetch.
The test doesn't have much value I agree but it is there the increase coverage (for the useEffect).

Reply:
> I don't see a lot of value mocking React.useState just to assert it was called
have in mind it is `setProjects` what's being asserted, not useState

Reply:
Yes, spied, sorry. I meant that just asserting `setProjects` was called, like without the projects, was a bit weak given the fact you are spying useState. Since you already have the fetch mocked, maybe asserting that it was called with the mocked projects would be more valuable.

Reply:
I disagree, we "don't care" about the fetch method here, only that the component is receiving projects at some point. Asserting `fetch` doesn't mean `setProjects` will be called, or that the projects will be added to the state.

In any case, the main reason why `setProjects` has to be asserted is because fetching is inside an async effect and it isn't possible to tell Enzyme to wait for that. The only way is to call `done` when `setProjects` is called.

Does this make sense?

Reply:
Well, I never said we should care about the fetch method, just that `setProjects` was receiving the right projects, which you actually are expecting here: https://github.com/3scale/porta/pull/1015/files#diff-9616eeee01eb7fb9cf1dbaf45ae410b9R63 which I missed in my first comment. I just find it a bit dirty, spying the setState, mocking the fetch and setProjects... but I guess it's ok if we can't just mock the fetch and assert after render the list items rendered or the props.

Reply:
OK all clear now. It maybe looks a bit hacky but it is the only way to do assertions with `useState` hook (as far as I know). So I will leave it like this if that's OK.